File: backend/chat-service/src/test/java/com/example/chatservice/api/internal/InternalApiFilterTest.java
package com.example.chatservice.api.internal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockFilterChain;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;

class InternalApiFilterTest {

    private final InternalApiFilter filter = new InternalApiFilter("secret-token");

    @Test
    void internalPathWithoutToken_isRejectedWith403() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-internal/chat/1");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();

        filter.doFilter(request, response, chain);

        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest(), "체인으로 넘어가면 안 된다");
    }

    @Test
    void internalPathWithWrongToken_isRejectedWith403() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-internal/chat/1");
        request.addHeader(InternalApiFilter.HEADER, "wrong");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();

        filter.doFilter(request, response, chain);

        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void internalPathWithCorrectToken_passesThrough() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("POST", "/api-internal/chat");
        request.addHeader(InternalApiFilter.HEADER, "secret-token");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();

        filter.doFilter(request, response, chain);

        assertEquals(200, response.getStatus());
        assertNotNull(chain.getRequest(), "체인으로 넘어가야 한다");
    }

    @Test
    void publicPath_isNotChecked() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-public/chat/1/rooms");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();

        filter.doFilter(request, response, chain);

        assertEquals(200, response.getStatus());
        assertNotNull(chain.getRequest());
    }

    @Test
    void debugPath_withoutToken_isRejectedWith403() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-debug/chat");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();

        filter.doFilter(request, response, chain);

        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void doubleSlashPrefix_isStillGuarded() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "//api-internal/x/1");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();
        filter.doFilter(request, response, chain);
        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void percentEncodedPrefix_isStillGuarded() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/%61pi-internal/x/1");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();
        filter.doFilter(request, response, chain);
        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void matrixParamInPrefix_isStillGuarded() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-internal;x=1/x/1");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();
        filter.doFilter(request, response, chain);
        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void dotDotTraversalFromPublic_isStillGuarded() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/api-public/../api-internal/x/1");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();
        filter.doFilter(request, response, chain);
        assertEquals(403, response.getStatus());
        assertNull(chain.getRequest());
    }

    @Test
    void debugPath_withCorrectToken_passesThrough() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest("DELETE", "/api-debug/x");
        request.addHeader(InternalApiFilter.HEADER, "secret-token");
        MockHttpServletResponse response = new MockHttpServletResponse();
        MockFilterChain chain = new MockFilterChain();
        filter.doFilter(request, response, chain);
        assertEquals(200, response.getStatus());
        assertNotNull(chain.getRequest());
    }

    @Test
    void blankToken_isRejectedAtConstruction() {
        org.junit.jupiter.api.Assertions.assertThrows(IllegalStateException.class, () -> new InternalApiFilter(" "));
    }
}

File: backend/chat-service/src/test/kotlin/com/example/chatservice/api/internal/InternalApiFilterTest.kt
package com.example.chatservice.api.internal

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.springframework.mock.web.MockFilterChain
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse

class InternalApiFilterTest {

    private val filter = InternalApiFilter("secret-token")

    private fun run(method: String, uri: String, token: String? = null): Pair<MockHttpServletResponse, MockFilterChain> {
        val request = MockHttpServletRequest(method, uri)
        if (token != null) request.addHeader(InternalApiFilter.HEADER, token)
        val response = MockHttpServletResponse()
        val chain = MockFilterChain()
        filter.doFilter(request, response, chain)
        return response to chain
    }

    @Test
    fun internalPathWithoutToken_isRejectedWith403() {
        val (response, chain) = run("GET", "/api-internal/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun internalPathWithWrongToken_isRejectedWith403() {
        val (response, chain) = run("GET", "/api-internal/x/1", "wrong")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun internalPathWithCorrectToken_passesThrough() {
        val (response, chain) = run("POST", "/api-internal/x", "secret-token")
        assertEquals(200, response.status)
        assertNotNull(chain.request)
    }

    @Test
    fun publicPath_isNotChecked() {
        val (response, chain) = run("GET", "/api-public/x/1")
        assertEquals(200, response.status)
        assertNotNull(chain.request)
    }

    @Test
    fun debugPath_withoutToken_isRejectedWith403() {
        val (response, chain) = run("GET", "/api-debug/x")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun doubleSlashPrefix_isStillGuarded() {
        val (response, chain) = run("GET", "//api-internal/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun percentEncodedPrefix_isStillGuarded() {
        val (response, chain) = run("GET", "/%61pi-internal/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun matrixParamInPrefix_isStillGuarded() {
        val (response, chain) = run("GET", "/api-internal;x=1/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun dotDotTraversalFromPublic_isStillGuarded() {
        val (response, chain) = run("GET", "/api-public/../api-internal/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun debugPath_withCorrectToken_passesThrough() {
        val (response, chain) = run("DELETE", "/api-debug/x", "secret-token")
        assertEquals(200, response.status)
        assertNotNull(chain.request)
    }

    @Test
    fun blankToken_isRejectedAtConstruction() {
        assertThrows(IllegalStateException::class.java) { InternalApiFilter(" ") }
    }

    @Test
    fun adminPath_withoutToken_isRejectedWith403() {
        val (response, chain) = run("GET", "/api-admin/x/1")
        assertEquals(403, response.status)
        assertNull(chain.request)
    }

    @Test
    fun adminPath_withCorrectToken_passesThrough() {
        val (_, chain) = run("GET", "/api-admin/x/1", "secret-token")
        assertNotNull(chain.request)
    }
}
